File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/EncodableParser.kt
package com.jetbrains.signatureverifier.serialization

import com.jetbrains.signatureverifier.serialization.dataholders.*
import org.bouncycastle.asn1.*


fun ASN1Primitive.toEncodableInfo(): EncodableInfo = when (this) {

  is ASN1TaggedObject -> TaggedObjectInfo(
    isExplicit,
    tagNo,
    this.baseObject.toASN1Primitive().toEncodableInfo()
  )

  is ASN1Sequence -> {
    SequenceInfo(this.map { it.toASN1Primitive().toEncodableInfo() })
  }

  is ASN1Set -> SetInfo(this.map { it.toASN1Primitive().toEncodableInfo() })

  else -> {
    try {
      val stringInfo = TextualInfo.getInstance(this)
      stringInfo
    } catch (_: IllegalArgumentException) {
      UnknownTypeInfo(this)
    }
  }
}

File: net/JetBrains.SignatureVerifier/src/Serialization/EncodableParser.cs
using System;
using System.Linq;
using Org.BouncyCastle.Asn1;

namespace JetBrains.SignatureVerifier.Serialization;

public static class EncodableParser
{
  public static IEncodableInfo ToEncodableInfo(this Asn1Object source)
  {
    switch (source)
       {
           case Asn1TaggedObject taggedObject:
               return new TaggedObjectInfo(
                   taggedObject.IsExplicit(),
                   taggedObject.TagNo,
                   taggedObject.GetObject().ToAsn1Object().ToEncodableInfo()
               );

           case Asn1Sequence sequence:
               return new SequenceInfo(sequence
                   .ToArray()
                   .Select(item => item.ToAsn1Object().ToEncodableInfo())
                   .ToList()
               );

           case Asn1Set set:
               return new SetInfo(set
                   .ToArray()
                   .Select(item => item.ToAsn1Object().ToEncodableInfo())
                   .ToList()
               );

           default:
               try
               {
                   return TextualInfo.GetInstance(source);
               }
               catch (Exception e)
               {
                   Console.WriteLine(e);
                   throw;
               }
       }
  }
}